### PR TITLE
gateway provisioner: support changing Kubernetes client log level

### PR DESCRIPTION
**Pull request by izturn**
add the flag `--kubernetes-debug` to enable Kubernetes client debug logging with log level.

Signed-off-by: Gang Liu [gang.liu@daocloud.io](mailto:gang.liu@daocloud.io)

### CODE REVIEW

File: apis/projectcontour/v1alpha1/contourdeployment.go [hunk 1]
```diff
@@ -59,6 +59,14 @@ type ContourSettings struct {
...
 	// +optional
 	NodePlacement *NodePlacement `json:"nodePlacement,omitempty"`
+
+	// KubernetesDebug Enable Kubernetes client debug logging with log level. If unset,
+	// defaults to 0.
+	//
+	// +kubebuilder:validation:Minimum=0
+	// +kubebuilder:validation:Maximum=9
+	// +optional
+	KubernetesDebug uint8 `json:"kubernetesDebug,omitempty"`
```
Comment (sunjayBhatia) [inline apis/projectcontour/v1alpha1/contourdeployment.go:69]: would prefer a rename to `KubernetesLogLevel` to make it a little clearer
```suggestion
	KubernetesLogLevel uint8 `json:"kubernetesDebug,omitempty"`
```

Comment (izturn) [reply apis/projectcontour/v1alpha1/contourdeployment.go:69]: just to let it match with
```
serve.Flag("kubernetes-debug", "Enable Kubernetes client debug logging with log level.").PlaceHolder("<log level>").UintVar(&ctx.KubernetesDebug)
```
if you think renaming it is better, I will change it

Comment (sunjayBhatia) [reply apis/projectcontour/v1alpha1/contourdeployment.go:69]: yeah I would prefer the rename: personally don't love the existing flag name and I would like to new field to be a bit more readable

no need to match the field name exactly with the flag, as people using the provisioner don't need to know the exact flag name, we can make it configurable with something more clear what the meaning is

Comment (izturn) [reply apis/projectcontour/v1alpha1/contourdeployment.go]: done

File: Makefile [hunk 2]
```diff
@@ -6,7 +6,7 @@ IMAGE := $(REGISTRY)/$(PROJECT)
 SRCDIRS := ./cmd ./internal ./apis
 LOCAL_BOOTSTRAP_CONFIG = localenvoyconfig.yaml
 SECURE_LOCAL_BOOTSTRAP_CONFIG = securelocalenvoyconfig.yaml
-ENVOY_IMAGE = docker.io/envoyproxy/envoy:v1.23.1
+ENVOY_IMAGE = docker.io/envoyproxy/envoy:v1.23.0
```
Comment (sunjayBhatia) [inline Makefile:9]: looks like a merge conflict, should be v1.23.1 here

File: changelogs/unreleased/4677-izturn-small.md [hunk 3]
```diff
@@ -0,0 +1 @@
+add the flag to enable Kubernetes client debug logging with log level.
```
Comment (sunjayBhatia) [inline changelogs/unreleased/4677-izturn-small.md:1]:
```suggestion
Add Kubernetes client debug log level configurability to ContourDeployment resource.
```

File: Makefile (hunk 2 above)
Comment (izturn) [reply Makefile]: done

File: changelogs/unreleased/4677-izturn-small.md (hunk 3 above)
Comment (izturn) [reply changelogs/unreleased/4677-izturn-small.md]: done

**Review by sunjayBhatia — COMMENTED** [review]
looks good! you'll have to run codegen again however

**Comment by izturn** [comment]
the  `make generate` OR `make generate-gateway-yam` is failed with 

> ...
> Generating Gateway API CRD YAML documents... 
> E0901 10:22:59.274967   49366 run.go:120] "command failed" err="git cmd = '/usr/bin/git fetch --depth=1 origin > v0.5.1-0.20220822185609-d05e07b14156': exit status 128"
> make: *** [Makefile:238：generate-gateway-yaml] 错误 1

any advice?

@sunjayBhatia

**Comment by sunjayBhatia** [comment]
> the `make generate` OR `make generate-gateway-yam` is failed with
> 
> > ...
> > Generating Gateway API CRD YAML documents...
> > E0901 10:22:59.274967   49366 run.go:120] "command failed" err="git cmd = '/usr/bin/git fetch --depth=1 origin > v0.5.1-0.20220822185609-d05e07b14156': exit status 128"
> > make: *** [Makefile:238：generate-gateway-yaml] 错误 1
> 
> any advice?
> 
> @sunjayBhatia

ah, should fix by merging `main` and when regenerating locally use `GATEWAY_API_VERSION=release-0.5 make generate`

we have a temporary bump to a branch not a tag of Gateway API to fix conformance tests that is causing that

File: internal/provisioner/model/model.go [hunk 4]
```diff
@@ -171,6 +171,14 @@ type ContourSpec struct {
 
 	// EnvoyWorkloadType is the way to deploy Envoy, either "DaemonSet" or "Deployment".
 	EnvoyWorkloadType WorkloadType
+
+	// KubernetesLogLevel Enable Kubernetes client debug logging with log level. If unset,
+	// defaults to 0.
+	//
+	// +kubebuilder:validation:Minimum=0
+	// +kubebuilder:validation:Maximum=9
+	// +optional
```
Comment (skriss) [inline internal/provisioner/model/model.go:180]: Don't need the kubebuilder markers here; if any other fields have them, they're just carried over from when this code was part of an API definition, but it is no longer so we don't need them.

File: internal/provisioner/objects/deployment/deployment.go [hunk 5]
```diff
@@ -97,7 +97,9 @@ func DesiredDeployment(contour *model.Contour, image string) *appsv1.Deployment
 		fmt.Sprintf("--contour-config-name=%s", contour.ContourConfigurationName()),
 		fmt.Sprintf("--leader-election-resource-name=%s", contour.LeaderElectionLeaseName()),
 		fmt.Sprintf("--envoy-service-name=%s", contour.EnvoyServiceName()),
+		fmt.Sprintf("--kubernetes-debug=%d", contour.Spec.KubernetesLogLevel),
```
Comment (skriss) [inline internal/provisioner/objects/deployment/deployment.go:100]: Please add some unit test cases for this.

**Review by skriss — CHANGES_REQUESTED** [review]
Sorry for the delayed review, just a couple small comments here.

File: internal/provisioner/model/model.go (hunk 4 above)
Comment (izturn) [reply internal/provisioner/model/model.go]: done

File: internal/provisioner/objects/deployment/deployment.go (hunk 5 above)
Comment (izturn) [reply internal/provisioner/objects/deployment/deployment.go:100]: done

File: internal/provisioner/model/model.go [hunk 6]
```diff
@@ -696,6 +692,7 @@ func New(cfg Config) *Contour {
 					},
 				},
 			},
+			KubernetesLogLevel: 1,
```
Comment (skriss) [inline internal/provisioner/model/model.go:695]: I think the default should stay as it was (i.e. 0)

Comment (izturn) [reply internal/provisioner/model/model.go:695]: i find the `New` only used in unittests, so set the `KubernetesLogLevel` to `1`, I will change it

Comment (skriss) [reply internal/provisioner/model/model.go]: Ah, I see. OK, let's get this in as-is, and maybe do some follow-up refactoring here. Thanks for the clarification!

**Review by skriss — APPROVED** [review]

**Pull request merged**